sdk/src/main/java/io/opentelemetry/sdk/common/export/CompletableResultCode.java

```diff
@@ -0,0 +1,99 @@
+package io.opentelemetry.sdk.common.export;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * The implementation of Export operations are often asynchronous in nature, hence the need to
+ * convey a result at a later time. CompletableResultCode facilitates this.
+ *
+ * <p>This class models JDK 8's CompletableFuture to afford migration should Open Telemetry's SDK
+ * select JDK 8 or greater as a baseline, and also to offer familiarity to developers.
+ */
+public class CompletableResultCode {
+  /** A convenience for declaring success */
+  public static final CompletableResultCode SUCCESS = new CompletableResultCode().success();
+
+  /** A convenience for declaring failure */
+  public static final CompletableResultCode FAILURE = new CompletableResultCode().failure();
+
+  public CompletableResultCode() {}
+
+  private volatile Boolean succeeded = null;
+  private final ArrayList<Runnable> actions = new ArrayList<>();
+  private final Lock state = new ReentrantLock();
+
+  /** The export operation finished successfully. */
+  public CompletableResultCode success() {
+    List<Runnable> actionsToRun = new ArrayList<>();
+    state.lock();
+    try {
+      if (succeeded == null) {
+        succeeded = true;
+        actionsToRun.addAll(actions);
```

Comment:
Do we need this copy? With `succeeded` always accessed in the lock seems ok without it

Reply:
Its paranoia on my part. By copying, we can avoid a deadlock on having this object's API called from within an action handler. Although... we're using a ReentrantLock so we can already avoid that situation! I shall remove the copying.

Reply:
Pushed changes